---
title: "SNPeff"
output: html_notebook
---

This documents using [SNPeff](https://pcingola.github.io/SnpEff/) to annotate SNPs in TanOak with their functional consequences.

## Install SNPeff

```{bash, eval=FALSE}
cd /usr/local/bin
wget https://snpeff.blob.core.windows.net/versions/snpEff_latest_core.zip
unzip snpEff_latest_core.zip
cp -s snpEff/snpEff.jar .
```


### Number of genes

```{bash}
cd "/Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared drives/TanOak/John_Maker/Round_2"

echo "gene: " $(grep -c "\sgene\s" TanOak2.all.gene.gff)

echo "CDS: " $(grep -c "\sCDS\s" TanOak2.all.gene.gff)

echo "mRNA: " $(grep -c "\smRNA\s" TanOak2.all.gene.gff)

echo "gene: " $(grep -c "\sgene\s" TanOak2.small.gff)

echo "CDS: " $(grep -c "\sCDS\s" TanOak2.small.gff)

echo "mRNA: " $(grep -c "\smRNA\s" TanOak2.small.gff)

```

## build genome index

set up directories and update snpEff.config file
```{bash, eval=FALSE}
echo "tanoak.v1221.genome : Tanoak" >> /usr/local/bin/snpEff/snpEff.config

mkdir /usr/local/bin/snpEff/data
mkdir /usr/local/bin/snpEff/data/tanoak.v1221
mkdir /usr/local/bin/snpEff/data/genomes
```

Move files
```{bash, eval=FALSE}
cd /usr/local/bin/snpEff/data/genomes

gzip -c /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly/PanGenome_ONT_pilon.purge.dedup90.drop2k.mergeN.fasta > tanoak.v1221.fa.gz

cd /usr/local/bin/snpEff/data/tanoak.v1221

gzip -c /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/John_Maker/Round_2/TanOak2.small.gff > genes.gff.gz

cp /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/John_Maker/Round_2/TanOak.transcripts.fa cds.fa

cp /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/John_Maker/Round_2/TanOak.proteins.fa protein.fa
```

build it
```{bash, eval=FALSE}
# DO NOT RUN THIS IN RSTUDIO
# RUN IN TERMINAL
cd /usr/local/bin/snpEff
/opt/homebrew/opt/openjdk/bin/java -jar snpEff.jar build -gff3 -v tanoak.v1221
```


check it
```{bash, eval=FALSE}
cd /usr/local/bin/snpEff
/opt/homebrew/opt/openjdk/bin/java -jar snpEff.jar dump tanoak.v1221   | head -n 40
```

## run annotation

```{bash, eval=FALSE}
cd ~/git/TanOak/output

time /opt/homebrew/opt/openjdk/bin/java -Xmx16g -jar /usr/local/bin/snpEff.jar tanoak.v1221 /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly/cohort.all.genotyped.vcf.gz > /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly/cohort.all.genotyped.snpEff.vcf

cd /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly
bcftools convert --output-type z cohort.all.genotyped.snpEff.vcf > cohort.all.genotyped.snpEff.vcf.bgz
rm cohort.all.genotyped.snpEff.vcf
tabix cohort.all.genotyped.snpEff.vcf.bgz

ln -s /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly/cohort.all.genotyped.snpEff.vcf.bgz ~/git/TanOak/output

ln -s /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly/cohort.all.genotyped.snpEff.vcf.bgz.tbi ~/git/TanOak/output

```



