---
title: "05 parse snpEff vcf"
output: html_notebook
---

The goal here is to find genes that have changes predicted to effect protein function and where those changes have different frequencies in the two classes.

## setup

```{r}
library(tidyverse)
library(VariantAnnotation)
datadir <- "~/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared drives/TanOak/new_12_2021_assembly/"
```

```{r}
dir(datadir)
```

Index the fasta file.  Only needs to be done once
```{bash, eval=FALSE}
samtools faidx "~/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared drives/TanOak/new_12_2021_assembly/PanGenome_ONT_pilon.purge.dedup90.drop2k.mergeN.fasta"
```

## filter

### filter to get a list of all SNPs that are homozygous ALT, so that we can update the fasta file to remove these

```{bash, eval=FALSE}
cd /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof\@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly/

module load bcftools # LSSC0

bcftools view --include 'COUNT(GT="1/1" | GT="1|1")=19'  --output-type z cohort.all.genotyped.snpEff.vcf.bgz > cohort.all.genotyped.snpEff.homozygous.alt.vcf.bgz
```

Convert it to a bed
```{bash, eval=FALSE}

module load bedops/2.4.33

bcftools view cohort.all.genotyped.snpEff.homozygous.alt.vcf.bgz | vcf2bed > cohort.all.genotyped.snpEff.homozygous.alt.bed
```

Filtering vcf to remove these:

```{bash, eval=FALSE}
bcftools view --exclude 'COUNT(GT="1/1" | GT="1|1")=19'  --output-type z cohort.all.genotyped.snpEff.vcf.bgz > cohort.all.genotyped.snpEff.no.homozygous.alt.vcf.bgz
```

Reality check:

```{bash, eval=FALSE}
bcftools view cohort.all.genotyped.snpEff.vcf.bgz | grep -v "#" | wc -l # 29947242
bcftools view  cohort.all.genotyped.snpEff.homozygous.alt.vcf.bgz | grep -v "#" | wc -l # 58314
bcftools view cohort.all.genotyped.snpEff.no.homozygous.alt.vcf.bgz | grep -v "#" | wc -l # 29888928
```

```{r}
29888928 + 58314 == 29947242
```

### filter to remove contigs that removed from v2 assembly (these are very small contigs)

```{bash, eval=FALSE}
bcftools view --targets "^102840_pilon,104688_pilon,114751_pilon,131185_pilon,116872_pilon,118222_pilon,122925_pilon,122931_pilon,200435_pilon,126988_pilon,131189_pilon,133381_pilon,133383_pilon,137344_pilon,140200_pilon,143367_pilon,161061_pilon,166030_pilon,167882_pilon,167884_pilon,169198_pilon,169990_pilon,181850_pilon,193062_pilon,174416_pilon,202968_pilon,202971_pilon,211916_pilon,183051_pilon,197829_pilon,208127_pilon,221338_pilon,266998_pilon,211908_pilon,226389_pilon,214831_pilon,240012_pilon,243729_pilon,246168_pilon,246076_pilon,251345_pilon,246422_pilon,246426_pilon,248231_pilon,251238_pilon,253755_pilon,264439_pilon,263762_pilon,85415_pilon,85416_pilon,86345_pilon,94891_pilon,142050_pilon,169158_pilon,193220_pilon,53324_pilon"  --output-type z cohort.all.genotyped.snpEff.no.homozygous.alt.vcf.bgz > cohort.all.genotyped.snpEff.no.homozygous.alt.v2.vcf.bgz

bcftools index --tbi cohort.all.genotyped.snpEff.no.homozygous.alt.v2.vcf.bgz
```


### for variants of moderate or high effect

```{bash, eval=FALSE}
cd /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof\@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly/

bcftools view --output-type z --include "INFO/ANN ~ 'HIGH' || INFO/ANN ~ 'MODERATE'" cohort.all.genotyped.snpEff.no.homozygous.alt.v2.vcf.bgz  > cohort.all.genotyped.snpEff.no.homozygous.alt.filtered_moderate_high.v2.vcf.bgz

bcftools index --tbi cohort.all.genotyped.snpEff.no.homozygous.alt.filtered_moderate_high.v2.vcf.bgz
```

### to remove intergenic SNPs

```{bash, eval=FALSE}
cd /Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof\@ucdavis.edu/Shared\ drives/TanOak/new_12_2021_assembly/

bcftools view --output-type z --exclude "INFO/ANN ~ 'intergenic_region'" cohort.all.genotyped.snpEff.no.homozygous.alt.v2.vcf.bgz > cohort.all.genotyped.snpEff.no.homozygous.alt.filtered_no_intergenic.v2.vcf.bgz

bcftools index --tbi cohort.all.genotyped.snpEff.no.homozygous.alt.filtered_no_intergenic.v2.vcf.bgz
```

# Filtering for High/Moderate

## make contig lengths file for those contigs remaining after moderate/high filtering

```{r}
contig.lengths <-
  system("bcftools view -h '/Users/jmaloof/Library/CloudStorage/GoogleDrive-jnmaloof@ucdavis.edu/Shared drives/TanOak/new_12_2021_assembly/cohort.all.genotyped.snpEff.no.homozygous.alt.filtered_moderate_high.v2.vcf.bgz' | grep contig", inter=TRUE) %>%
  as_tibble() %>%
  mutate(value=str_remove_all(value, "(##contig=<ID=)|(length=)|>")) %>%
  separate(value, into = c("name", "length"), sep=",", convert = TRUE )
contig.lengths
```


## take a look at a single contig

load it
```{r}
tab <- TabixFile(file.path(datadir, "cohort.all.genotyped.snpEff.no.homozygous.alt.filtered_moderate_high.v2.vcf.bgz"))
vcfVR <- readVcfAsVRanges(tab, param=GRanges(seqnames = contig.lengths$name[1],
                                  ranges = IRanges(start = 1, end=contig.lengths$length[1])))
vcf <-  readVcf(tab, param=GRanges(seqnames = contig.lengths$name[1],
                                  ranges = IRanges(start = 1, end=contig.lengths$length[1])))
```

### quality
```{r}
tibble(QUAL=rowRanges(vcf)$QUAL) %>% 
  ggplot(aes(x=QUAL)) +
  geom_histogram(fill="skyblue") +
  scale_x_log10() +
  ggtitle("Quality") +
  xlab("QUAL") +
  ylab("Count")
min(rowRanges(vcf)$QUAL)
```

### Genotyped allele number
This is the number of allele genotype calls made.  Because there are multipe preps of some samples, there are a total of 19 samples, max AN is 38.  It might be nice to filter this for the actual samples we are going to analyze, but for now just filter for 38
```{r}
info(vcf)$AN %>% tibble(AN=.) %>%
  ggplot(aes(x=AN)) +
  geom_histogram(fill="red", binwidth = 1) +
  ggtitle("Number of Alleles Called") +
  xlab("Allele Number (AN)") +
  ylab("Count") +
  scale_x_continuous(breaks=seq(2,40,2))
```

### Number of alternative alleles
"1" means a b
```{r}
sapply(info(vcf)$AC, length) %>% tibble(AC=.) %>%
  ggplot(aes(x=AC)) +
  geom_histogram(fill="springgreen2", binwidth = 1) +
  ggtitle("Number of Alternate Alleles") +
  xlab("Alternative alleles") +
  ylab("Count") +
  scale_x_continuous(breaks=seq(1,6,1))
```

Overall this looks like little filtering is needed.  But let's do a little.

```{r}
vcf.filter <- vcf[rowRanges(vcf)$QUAL > 50 &
                    sapply(info(vcf)$AC, length) == 1 & #stick with biallelic SNPs for now
                    info(vcf)$AN == 38] # require complete genotype info for now 
length(vcf)
length(vcf.filter)
```

Could do a VRanges object and convert to tibble:
```{r}
system.time(
  vcf.filter.VR <- as(vcf.filter, "VRanges") 
)
```


```{r}
system.time(
  VRtibble <- tibble(
                 seqname=as.character(seqnames(vcf.filter.VR)),
                 start=start(vcf.filter.VR),
                 end=end(vcf.filter.VR),
                 sample=as.character(sampleNames(vcf.filter.VR)),
                 ref=ref(vcf.filter.VR),
                 alt=alt(vcf.filter.VR),
                 RD=refDepth(vcf.filter.VR),
                 AD=altDepth(vcf.filter.VR),
                 as.data.frame(mcols(vcf.filter.VR)[c("QUAL", "GT", "GQ", "PL", "ANN" )])
                 ) %>%
    mutate(snpID=str_c(seqname,start,sep=":"),
           GT=str_replace(GT, stringr::fixed("|"), "/")) 
) #6 seconds

```

reformat the snpEff annotation into tibble form
```{r}
annnames <- c("allele", "effect", "impact", "GeneName", "GeneID", "FeatureType", "FeatureID", "TranscriptType", "Rank_Total", "HGVS.c", "HGVS.p", "cDNApos_len", "CDSpos_len", "protpos_len", "distance", "warn")

VRtibble <- VRtibble %>% separate(ANN, sep = ",", into=c("ANN1","ANN2", "ANN3", "ANN4") ) %>%
  pivot_longer(cols = starts_with("ANN"), names_to = "ANN_ID", values_to = "ANN", values_drop_na = TRUE) %>%
  separate(ANN, into=annnames, sep = "\\|") %>%
filter(impact != "MODIFIER")

VRtibble
```


## Can I do all at once, since I have filtered this down to a reasonable size?


```{r}
tab <- TabixFile(file.path(datadir, "cohort.all.genotyped.snpEff.no.homozygous.alt.filtered_moderate_high.v2.vcf.bgz"))

system.time(
  vcf <-  readVcf(tab)
  )
```


### quality
```{r}
tibble(QUAL=rowRanges(vcf)$QUAL) %>% 
  ggplot(aes(x=QUAL)) +
  geom_histogram(fill="skyblue") +
  scale_x_log10() +
  ggtitle("Quality") +
  xlab("QUAL") +
  ylab("Count")
min(rowRanges(vcf)$QUAL)
```

### Genotyped allele number
This is the number of allele genotype calls made.  With 14 samples, the max here is 28
```{r}
info(vcf)$AN %>% tibble(AN=.) %>%
  ggplot(aes(x=AN)) +
  geom_histogram(fill="red", binwidth = 1) +
  ggtitle("Number of Alleles Called") +
  xlab("Allele Number (AN)") +
  ylab("Count") +
  scale_x_continuous(breaks=seq(2,40,2))
```

### Number of alternative alleles
"1" means a b
```{r}
sapply(info(vcf)$AC, length) %>% tibble(AC=.) %>%
  ggplot(aes(x=AC)) +
  geom_histogram(fill="springgreen2", binwidth = 1) +
  ggtitle("Number of Alternate Alleles") +
  xlab("Alternative alleles") +
  ylab("Count") +
  scale_x_continuous(breaks=seq(1,6,1))
```

Overall this looks like little filtering is needed.  But let's do a little.

```{r}
vcf.filter <- vcf[rowRanges(vcf)$QUAL > 50 &
                    sapply(info(vcf)$AC, length) == 1 & #stick with biallelic SNPs for now
                    info(vcf)$AN == 38] # require complete genotype info for now 
length(vcf)
length(vcf.filter)
```

Could do a VRanges object and convert to tibble:
```{r}
system.time(
  vcf.filter.VR <- as(vcf.filter, "VRanges") 
)
```


```{r}
system.time(
  VRtibble <- tibble(
                 seqname=as.character(seqnames(vcf.filter.VR)),
                 start=start(vcf.filter.VR),
                 end=end(vcf.filter.VR),
                 sample=as.character(sampleNames(vcf.filter.VR)),
                 ref=ref(vcf.filter.VR),
                 alt=alt(vcf.filter.VR),
                 RD=refDepth(vcf.filter.VR),
                 AD=altDepth(vcf.filter.VR),
                 as.data.frame(mcols(vcf.filter.VR)[c("QUAL", "GT", "GQ", "PL", "ANN" )])
                 ) %>%
    mutate(snpID=str_c(seqname,start,sep=":"),
           GT=str_replace(GT, stringr::fixed("|"), "/")) 
) #35 seconds

```

reformat the snpEff annotation into tibble form
```{r}
annnames <- c("allele", "effect", "impact", "GeneName", "GeneID", "FeatureType", "FeatureID", "TranscriptType", "Rank_Total", "HGVS.c", "HGVS.p", "cDNApos_len", "CDSpos_len", "protpos_len", "distance", "warn")

system.time(
  VRtibble <- VRtibble %>% separate(ANN, sep = ",", into=c("ANN1","ANN2", "ANN3", "ANN4", "ANN5", "ANN6") ) %>%
  pivot_longer(cols = starts_with("ANN"), names_to = "ANN_ID", values_to = "ANN", values_drop_na = TRUE) %>%
  separate(ANN, into=annnames, sep = "\\|") %>%
filter(impact != "MODIFIER")
)

head(VRtibble)
```
summary by effect:

```{r}
table(VRtibble$effect, VRtibble$impact) %>%
  as.data.frame() %>%
  dplyr::rename(effect=Var1, impact=Var2, freq=Freq) %>%
  pivot_wider(names_from = "impact", values_from = "freq") %>%
  dplyr::select(effect, HIGH, MODERATE, LOW) %>%
  arrange(desc(HIGH), desc(MODERATE), desc(LOW)) %>%
  write_csv("../output/impact_high_mod.csv")
```


```{r}
save(VRtibble, file=file.path(datadir, "SnpEffFilteredTibble.Rdata"))
```

