---
title: "PCA of SNPs"
output: html_notebook
---

Will make a PCA plot of SNPs, colored by phenotype.

In this version, sample 9 times to test whether we are sampling enough SNPs.

```{r}
library(tidyverse)
library(VariantAnnotation)
datadir <- "/Volumes/GoogleDrive/Shared drives/TanOak/"
outdir <- "../output/"
```

```{r}
tab <- TabixFile(file.path(datadir, "cohort.genotyped.snpEff.vcf.bgz"))

tab_seqnames <- seqnamesTabix(tab)

## load contig lengths file 
contig.lengths <- read_csv("../output/contigLengths.csv")

contig.lengths_tab <- contig.lengths %>%
  filter(name %in% tab_seqnames)
```

```{r}
## get pheno data
pheno <- read_csv("../input/TanOakResistance.csv") %>%
  mutate(sampleID=str_replace_all(Tanoak, "-","."),
         sampleID=str_replace(sampleID, "SM\\.(52|54)\\.(42|81|97|37)", "\\1.\\2")) %>%
  arrange(Rank_Sorting)
```

tibble to hold results
```{r}
results <- tibble(trial=1:9)
```

function to generate PCs
```{r}
getPCs <- function() {
  
  #select a subset of scaffolds to load
  contigs.subset <- sample(1:nrow(contig.lengths_tab), size = 200)
  
  #load them
  vcf <-  readVcf(tab, 
                  param=GRanges(
                    seqnames = contig.lengths_tab$name[contigs.subset],
                    ranges = IRanges(
                      start = 1,
                      end=contig.lengths_tab$length[contigs.subset])))
  
  # filter it
  vcf <- vcf[rowRanges(vcf)$QUAL > 50 &
               sapply(info(vcf)$AC, length) == 1 & #stick with biallelic SNPs for now
               info(vcf)$AN == 28] # require complete genotype info for now 
  
  mat <- genotypeToSnpMatrix(vcf)
  mat3 <- as(mat$genotypes, "numeric")
  
  mat3 <- mat3[,apply(mat3, 2, function(x) !any(is.na(x)))]
  
  
  geno.pca <- prcomp(mat3, rank.=10)
  
  return(geno.pca)
}
```

function to calculate pct var
```{r}
getPctvar <- function(geno.pca) {
  pcvar <- geno.pca$sdev^2 # square std dev to get variance
  pcvar.pct <- tibble(pctvar=pcvar/sum(pcvar) * 100,
                      PC=1:length(pcvar))
}
```

function to make PCA frame for plotting
```{r}
getPCtibble <- function(geno.pca) {
PCs <- geno.pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sampleID") %>%
  left_join(pheno)
PCs
}
```

run it
```{r}
results <- results %>%
  mutate(PCs = map(trial, ~ getPCs())) %>%
  mutate(pctvar = map(PCs, getPctvar)) %>%
  mutate(PCtibble = map(PCs, getPCtibble))
gc()
```

```{r}
results %>% dplyr::select(trial, pctvar) %>% 
  unnest(pctvar) %>%
  ggplot(aes(x=PC, y=pctvar, col=as.character(trial))) +
  geom_col(position = "dodge")
```

```{r, fig.asp=1, fig.height=10}
results %>% dplyr::select(trial, PCtibble) %>% 
  unnest(PCtibble) %>%
  ggplot(aes(x=PC1, y=PC2, color=Tolerance_Resistance, label=sampleID)) +
  geom_text(size=3) +
  facet_wrap(~ trial, scales = "free") +
  theme(legend.position = "top")
```


```{r}
PCs %>%
  ggplot(aes(x=PC1, y=PC2, color=Tolerance_Resistance, label=sampleID)) +
  geom_text() 
```


