---
title: "09_Explore_dN_dS"
author: "Julin Maloof"
date: 'updated 2023-04-24'
output: html_document
---

Goal is to look at the genes in the  GO:0043565; sequence-specific DNA binding category with high dNdS.
* What are their predicted functions
* SNP polarity (does Tanoak or Oak have the ancestral alleles?)

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
library(org.At.tair.db)
library(Biostrings)
```

## get data

dn_ds data
```{r}
dn_ds <- read_csv("../output/gene_sets_dnds_long.csv")
head(dn_ds)
```

annotation data
```{r}
Atblast <- read_csv("../output/TanOakVsA.t.blast_out.csv", col_names = str_split("genes AGI pident length mismatch gapopen qstart qend sstart send evalue bitscore", " ")[[1]])
head(Atblast)
```


```{r}
interpro <- read_tsv("../output/12_21_tanoak_interpro.tsv.gz", col_names = FALSE)
#interpro <- read_tsv("../output/output.iprscan", col_names = FALSE)
head(interpro)
```


## DNA Binding

Get the genes
```{r}
DNA_binding_genes <- dn_ds %>% filter(all_go_terms=="GO:0043565", dnds >= 1.4)
DNA_binding_genes
```

interpro annotation

```{r}
dna.interpro <- interpro %>% 
  dplyr::select(genes=X1, description=X6 ) %>% 
  mutate(genes=str_remove(genes, "-mRNA-.*")) %>%
  unique() %>% 
  na.omit() %>% 
  right_join(DNA_binding_genes, by="genes") %>%
  group_by(genes) %>%
  summarize(interpro.description = str_c(description, collapse="; "), dnds=unique(dnds))
  
dna.interpro
```

Arabidopsis best hit and gene symbols (if any)

```{r}
dna.at <- Atblast %>%
  dplyr::select(genes, AGI, pident, length, evalue, bitscore) %>%
  mutate(genes=str_remove(genes, "-mRNA-.*")) %>%
  right_join(dna.interpro) %>%
  group_by(genes) %>%
  slice_max(order_by=bitscore, n=1) %>%
  mutate(AGI=str_remove(AGI, "\\.[0-9]"),
         AT.symbol = unlist(as.list(org.At.tairSYMBOL[AGI]))[[1]]) %>%
  dplyr::select(genes, AT.symbol, AGI, interpro.description, everything()) %>%
  arrange(desc(dnds))

dna.at
```

### Notes:

__RAS1__ RESPONSE TO ABA AND SALT 1;(source:Araport11)

__ATHB-7__ encodes a putative transcription factor that contains a homeodomain closely linked to a leucine zipper motif. Transcript is detected in all tissues examined. Is transcriptionally regulated in an ABA-dependent manner and may act in a signal transduction pathway which mediates a drought response.

__ATWRKY40__  	Pathogen-induced transcription factor. Binds W-box sequences in vitro. Forms protein complexes with itself and with WRKY40 and WRKY60. Coexpression with WRKY18 or WRKY60 made plants more susceptible to both P. syringae and B. cinerea. WRKY18, WRKY40, and WRKY60 have partially redundant roles in response to the hemibiotrophic bacterial pathogen Pseudomonas syringae and the necrotrophic fungal pathogen Botrytis cinerea, with WRKY18 playing a more important role than the other two. The mRNA is cell-to-cell mobile.

__ATWRKY41__ member of WRKY Transcription Factor; Group III; AvrRps4C can associate with the DNA-binding domain.

__ATWRKY11__ member of WRKY Transcription Factor; Group II-d; negative regulator of basal resistance to Pseudomonas syringae.

__ATHSFA2__ member of Heat Stress Transcription Factor (Hsf) family. Involved in response to misfolded protein accumulation in the cytosol. Regulated by alternative splicing and non-sense-mediated decay.

__AT-HSFA3__ Member of Heat Stress Transcription Factor (Hsf) family. Expression is regulated by DREB2A and in turn HSFA3 regulates the expression of hsps Hsp18.1-CI and Hsp26.5-MII35S. Involved in establishing thermotolerence.



Annotate All    NOT WORKING...NEED TO REMOVE REDUNDANCY /PIVOT WIDER AND COMBINE COLUMNS, OR SOMETHING
```{r}
all.interpro <- interpro %>% 
  dplyr::select(genes=X1, description=X6 ) %>% 
  mutate(genes=str_remove(genes, "-mRNA-.*")) %>%
  unique() %>% 
  na.omit() %>% 
  right_join(dn_ds, by="genes") %>%
  group_by(genes) %>%
  summarize(interpro.description = str_c(description, collapse="; "), dnds=unique(dnds))
  
dna.interpro

all.at <- Atblast %>%
  dplyr::select(genes, AGI, pident, length, evalue, bitscore) %>%
  mutate(genes=str_remove(genes, "-mRNA-.*")) %>%
  right_join(interpro) %>%
  group_by(genes) %>%
  slice_max(order_by=bitscore, n=1) %>%
  mutate(AGI=str_remove(AGI, "\\.[0-9]"),
         AT.symbol = unlist(as.list(org.At.tairSYMBOL[AGI]))[[1]]) %>%
  dplyr::select(genes, interpro.description, AGI, AT.symbol, everything()) %>%
  arrange(desc(dnds))

all.at
```


Let's blast a few of these.  First create fasta file of interesting proteins:

```{r}
peps <- readAAStringSet("../input/blast/TanOak.proteins.fa")
names(peps) <- names(peps) %>% str_remove("-mRNA-.*")
peps <- peps[DNA_binding_genes$genes]
peps
writeXStringSet(peps, "../output/DNA_binding_peps.fa")
```


