---
title: "Explore DNA Binding"
output: html_notebook
---

The GO category DNA binding is enriched for high dNdS genes.   The goal here is to align the TanOak genes with other related homologs, build a tree, and see where changes are happening.

Previously I blasted against the NR database, and that was a bit of a mess for building a tree.  Will try again, but with BLAST against selected organisms.

```{r}
library(tidyverse)
library(Biostrings)
library(UpSetR)
library(gplots)
```


## Homolog Sequence retrieval

Start with blastp of `../output/DNA_binding_peps.fa`

Used phyotozome blastp tool against: 

* Athaliana_TAIR10_167.proteome; 
* Qrubra_v2_1_687.proteome;
* Cdentata_v1_1_673.proteome; 
* Gmax_Wm82_a4_v1_508.proteome;
* Ptrichocarpa_v4_1_533.proteome


Results table, retrieved peptides and cDNAs are in `output/DNA_binding_downloads`

## Sequence overlap

Have common sequences been retrieved for the different queries?

```{r}
hittable <- read_tsv("../output/DNA_binding_downloads/DNA_Binding_Blast.tsv",
                     name_repair = make.names) %>%
  select(-1) %>%
  mutate(gene.name = str_remove(Protein, "\\.[0-9]{1,2}(\\.p)?$"),
         Protein = str_remove(Protein, "\\.p$")) %>% # to match fasta file
  group_by(Query.ID, gene.name) %>%
  slice_max(Bitscore) %>% # one best hit per Query/ Subject
  ungroup()

hittable
```

```{r, fig.width=8}
hit.counts <- hittable %>% select(Protein, Query.ID) %>%
  unique() %>%
  table() %>%
  as.data.frame.matrix() 

upset(hit.counts, nsets = 10, nintersects = 50)
```

## load Tanoak DNA_binding sequences
```{r}
tanoak_dna_bind_aa <- readAAStringSet("../output/DNA_binding_peps.fa")
```

## load all Tanoak sequences:

```{r}
tanoak_aa <- readAAStringSet("../input/blast/TanOak.proteins.fa")
names(tanoak_aa) <- names(tanoak_aa) %>% str_remove("-mRNA-.*")
```


## Load blast sequences

```{r}
#blast results
blast_peps <- readAAStringSet("../output/DNA_binding_downloads/DNA_Binding_Blast.fas.txt")
names(blast_peps) <- names(blast_peps) %>% str_remove(".*\\|") # get it to match hittable$Protein
```

## original oak Qrob orthologs

```{r}
blastp_ortho <- read_csv("../output/tanoak_oak_blastp_orthologs.csv") %>%
  mutate(query_id=str_remove(query_id, "-mRNA-.*"))
#read the Oak AA sequences and fix names
qrobAA <- readAAStringSet("../input/blast/Qrob_PM1N_CDS_aa_20161004.fa")
names(qrobAA) <- str_remove(names(qrobAA), " .*") #get rid of extra space and numbers
```

## Qrob homologs

Do blast to get all Qrob homologs

```{bash, eval=FALSE}
cd ../input/blast
blastp -query ../../output/DNA_binding_peps.fa -db Qrob_PM1N_CDS_aa_20161004.fa -outfmt 7 > ../../output/DNA_binding_Qrob_blast.tsv
```

```{r}
header <- str_split("query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score", pattern = ", ") %>%
  unlist() %>%
  make.names() %>%
  str_replace(fixed(".."), ".") %>%
  str_replace("X", "pct")

Qrob_hittable <- read_tsv("../output/DNA_binding_Qrob_blast.tsv", comment = "#", col_names = header)

Qrob_hittable
```

## Make a function for extracting homologs and building a tree

```{r}

get_homologs_tree <- function(ids, filestem, my.hittable=hittable, my.blast_peps=blast_peps, my.blastp_ortho=blastp_ortho, my.Qrob_hittable=Qrob_hittable, my.qrobAA=qrobAA, my.tanoak_aa=tanoak_aa, my.tanoak_dna_bind_aa=tanoak_dna_bind_aa) {
  
  
  # get the blast hits
  hits <- my.hittable %>%
    filter(Query.ID %in% ids) %>%
    filter(!duplicated(gene.name))
  
  homologs <- my.blast_peps[hits$Protein]
  
  #add tanoak
  
  targets <- my.tanoak_dna_bind_aa[ids]
  
  names(targets) <- names(targets) %>% str_c("target_", .)
  homologs <- c(targets, homologs)
  
  # get the oak protein orthologs
  
  # filter for those that we want
  oak_orthos <- my.blastp_ortho %>%
    filter(query_id %in% ids)
  
  qrobAA_orthos <- my.qrobAA[oak_orthos$subject_id]
  
  #add "ortho" to name to make it clear
  names(qrobAA_orthos) <- str_c(names(qrobAA_orthos), "_ortho")
  
  #add oak
  homologs <- c(qrobAA_orthos, homologs)
  
  
  # now add other Qrob homologs:
  
  Qrob_hits <- Qrob_hittable %>%
    filter(query.acc.ver %in% ids,
           !duplicated(subject.acc.ver),
           bit.score >= 100) %>%
    pull(subject.acc.ver)
  
  qrob_homologs <- my.qrobAA[Qrob_hits]
  
  homologs <- c(homologs, qrob_homologs)
  
  # Other Tanoak homologs
  
  if(!dir.exists("_tmp")) dir.create("_tmp")
  
  writeXStringSet(my.tanoak_dna_bind_aa[ids], "_tmp/tanoak_tmp_aa.fa")
  
  system("blastp -query _tmp/tanoak_tmp_aa.fa -db ../input/blast/TanOak.proteins.fa -outfmt 7 > _tmp/tmp_vs_Tanoak_blastp.tsv")
  
  hits_tanoak <- read_tsv("_tmp/tmp_vs_Tanoak_blastp.tsv",
                          comment = "#",
                          col_names = header) %>%
    filter(evalue < 1e-5,
           !subject.acc.ver %in% unique(query.acc.ver),
           !duplicated(subject.acc.ver)) %>%
    mutate(subject.acc.ver = str_remove(subject.acc.ver, "-mRNA-.*"))
  
  # extract the TanOak sequences and add
  
  tanoak_paralogs <- my.tanoak_aa[hits_tanoak$subject.acc.ver]
  homologs <- c(homologs, tanoak_paralogs)
  
  # Prune short sequences
  med_length <- median(nchar(homologs))
  homologs <- homologs[nchar(homologs) > 0.5*med_length & nchar(homologs) < 2*med_length]
  
  
  # write out the table
  names(homologs) <- str_remove(names(homologs), ".*\\|")
  homologs <- homologs[!duplicated(names(homologs))]
  writeXStringSet(homologs, "_tmp/tmp_tanoak_orthos.fa")
  
  ## align!
  if(!(dir.exists("trees"))) dir.create("trees")
  system2("mafft",
          args = c("--thread 4", "--localpair", "--maxiterate 1000", "--reorder", "_tmp/tmp_tanoak_orthos.fa"),
          stdout=str_c("../output/trees/", filestem, "_orthos_aligned.fa"))
  
  # prune alignment
  
  aligned <- readAAMultipleAlignment(str_c("../output/trees/", filestem, "_orthos_aligned.fa"))
  aligned <- maskGaps(aligned, min.fraction=0.50, min.block.width=10)
  maskedratio(aligned)
  
  writeXStringSet(as(aligned, "AAStringSet"), str_c("../output/trees/", filestem, "_orthos_aligned_pruned.fa"))
  
  # build tree
 # system2("FastTreeMP", args=str_c("../output/trees/", filestem, "_orthos_aligned_pruned.fa"),
          #stdout = str_c("../output/trees/", filestem, "_orthos.tre"))

system2("FastTree", args=str_c("../output/trees/", filestem, "_orthos_aligned_pruned.fa"),
          stdout = str_c("../output/trees/", filestem, "_orthos.tre"))

# get close orthologs from oak and chestnut

ruba_dentata_ortho_names <- hits %>% filter(Species %in% c("C.dentata v1.1", "Q.rubra v2.1") ) %>%
  group_by(Species) %>%
  slice_max(Bitscore) %>% 
  pull(Protein)

orthos_small <- c(targets, qrobAA_orthos, homologs[ruba_dentata_ortho_names])

writeXStringSet(orthos_small, "_tmp/tmp_tanoak_small_orthos.fa")

system2("mafft",
        args = c("--thread 4", "--localpair", "--maxiterate 1000", "--reorder", "_tmp/tmp_tanoak_small_orthos.fa"),
        stdout=str_c("../output/trees/", filestem, "_small_orthos_aligned.fa"))

file.remove(dir("_tmp", pattern=".*", full.names = TRUE))
}
```


```{r}
get_homologs_tree(ids = c("augustus-1014_pilon-processed-gene-5.8",
                          "augustus-542_pilon-processed-gene-26.72"),
                  filestem="1014_542_RAS1")

get_homologs_tree(ids = c("maker-431_pilon-augustus-gene-9.19", "augustus-266_pilon-processed-gene-97.305"), filestem = "431_266_ATHB-7")

get_homologs_tree(ids = "maker-74056_pilon-snap-gene-18.97",
                  filestem = "74056_WRK41")

get_homologs_tree(ids = "maker-261_pilon-snap-gene-47.3",
                  filestem = "261_WRKY40")

get_homologs_tree(ids = "maker-345_pilon-snap-gene-67.1",
                  filestem = "345_WRKY11")

get_homologs_tree(ids = c("augustus-353_pilon-processed-gene-14.64", "augustus-267_pilon-processed-gene-29.9"),
                    filestem = "353_HSF2_3")
```

See trees on iTol

STOPPED HERE

## Export Tanoak, oak, and chestnut sequences:

```{r}
#585
homologs_804_585[c("augustus_masked-585_pilon-processed-gene-3.11-mRNA-1", "Qrob_P0680550.2", "Caden.12G048400", )] %>%
  writeXStringSet("../output_585_orthos_small.fa")

#804
homologs_804_585[c("augustus_masked-804_pilon-processed-gene-0.1-mRNA-1","Qrob_P0468790.2", "Caden.12G048800")] %>%
  writeXStringSet("../output_804_orthos_small.fa")
```

```{bash}
time mafft --thread 3 --localpair --maxiterate 1000 ../output_585_orthos_small.fa > ../output_585_orthos_small_aligned.fa
```

```{bash}
time mafft --thread 3 --localpair --maxiterate 1000 ../output_804_orthos_small.fa > ../output_804_orthos_small_aligned.fa
```

